protocol/react/README.md

```diff
@@ -0,0 +1,121 @@
+React Protocol
+==============
+
+A guide for writing great front-end web apps.
+
+General Philosophies
+--------------------
+
+* For greenfield React projects we like to use TypeScript. TypeScript is a
+  typed superset of JavaScript that compiles to plain JavaScript. For a quick
+introduction, check out [TypeScript in 5
+minutes](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html).
+* We prefer to use [Function
+  Components](https://reactjs.org/docs/components-and-props.html#function-and-class-components)
+over Class components whenever possible. That means, as a starting point
+for state management, we should be using
+[react-hooks](https://reactjs.org/docs/hooks-intro.html) specifically
+`useState`.
+* For side-effects and lifecycle methods, use `useEffect` instead of a class
+  component.
+* If our app needs to make network requests and hold state outside of a
+  component, we like using GraphQL and our preferred library is
+[Apollo](https://www.apollographql.com). We have a [section on
+`GraphQL`](https://github.com/thoughtbot/guides/blob/master/graphql/README.md)
+in our Guides.
+* When building React apps with TypeScript and Apollo, we've found working
+  in [VSCode](https://code.visualstudio.com/) to be a mostly-good experience.
+
+Set Up Laptop
+-------------
+
+Set up your laptop with [this script](https://github.com/thoughtbot/laptop) and
+[these dotfiles](https://github.com/thoughtbot/dotfiles).
+
+Create App
+----------
+
+In the terminal, initialize the app with
+[create-react-app](https://github.com/facebook/create-react-app) and TypeScript.
+`create-react-app` is an open source React boilerplate tool similar to
+Thoughtbot's own Suspenders for Rails. It creates a basic React app with no
+build configuration and a good set of defaults.
+
+    yarn create react-app app-name --typescript
+
+Set Up App
+----------
+
+We're huge fans of automatic code formatting and linting.
+[Prettier](https://prettier.io) and [ESLint](https://eslint.org/) are both great
+libraries to facilitate this.
+
+Add `eslint` to the project
+
+    yarn add eslint @typescript-eslint/parser @typescript-eslint/eslint-plugin --dev
+
+The following is a `.eslintrc.js` file that has some sensible defaults. Add this
+to the root of your project.
+
+<details>
```

Comment:
Typically you'll want to include the summary in `<summary>` element, and that will be rendered as the text with the disclosure triangle (rather than the generic `"Details"`).

The following example shows using it w/ a JS formatted code block (view the raw comment to see the specific markup as it is a bit tricky).

<details>
  <summary>Sample <code>.eslintrc.js</code> file with some sensible defaults</summary>

``` js
module.exports = {
  end: {
    node: true,
    es6: true,
  }
}
```
</details>